django/middleware/csrf.py

```diff
@@ -155,9 +158,29 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
                 if referer is None:
                     return self._reject(request, REASON_NO_REFERER)
 
-                # Note that request.get_host() includes the port.
-                good_referer = 'https://%s/' % request.get_host()
-                if not same_origin(referer, good_referer):
+                referer = urlparse(referer)
+
+                # Make sure we have a valid URL for Referer
+                if '' in (referer.scheme, referer.netloc):
+                    return self._reject(request, REASON_MALFORMED_REFERER)
+
+                # Ensure that our Referer is also secure
+                if referer.scheme != 'https':
+                    return self._reject(request, REASON_INSECURE_REFERER)
+
+                # If there is no CSRF_COOKIE_DOMAIN, assume we need
+                # an exact match on host:port.
+                # if not, obey the cookie rules.
+                if settings.CSRF_COOKIE_DOMAIN is None:
+                    # Note that request.get_host() includes the port.
+                    good_referer = request.get_host()
+                else:
+                    good_referer = settings.CSRF_COOKIE_DOMAIN
+                    server_port = request.META['SERVER_PORT']
+                    if server_port != 443:
+                        good_referer = '%s:%d' % (good_referer, server_port)
```

Comment:
And lastly the builtin wsgi python server: https://github.com/python/cpython/blob/aa8ea3a6be22c92e774df90c6a6ee697915ca8ec/Lib/wsgiref/simple_server.py#L58